### ISSUE TITLE
hterm

### ISSUE THREAD

**Issue by dprutean** [jline/jline3#606]

**Comment by gnodet** [comment]
If you can somehow create an [ExternalTerminal](https://github.com/jline/jline3/blob/master/terminal/src/main/java/org/jline/terminal/impl/ExternalTerminal.java), you should be ok.

**Comment by dprutean** [comment]
Thank you for your feedback. I succeeded to create ExternalTerminal, but still, some code above does not work as expected. I have updated the issue text.

**Comment by gnodet** [comment]
`Terminal.TYPE_DUMB` is certainly now what you want, you should try with "xterm".  
I suppose an exception is thrown by the `readLine`  call, can you paste the stack trace ?

**Comment by dprutean** [comment]
There is no exception, but the readLine() never returns.

**Comment by gnodet** [comment]
Yes, that's certainly the problem, try translating those to `\n` instead.

**Comment by wise-coders** [comment]
Now I tried this but still does not help. I think I miss something else.

**Comment by wise-coders** [comment]
I found the problem. The inputStream class has to implement also ```public synchronized int read(byte b[], int off, int len).```

But the autocompletion is not working, as well as the text highlighting. How may I activate them? Does the external terminal need more specifications? Should I supply something else than 'xterm' in the ExternalTerminal constructor?

**Comment by gnodet** [comment]
If you use the stream you outlined above, not much can work, as JLine will only deal with a mostly complete line.  For completion to work, you can't buffer line by lines using the `commandQueue` and you need to put characters in the stream as soon as possible.
In the initial snippet, you were using `Type.DUMB` which will disable a lot of features, including colouring, etc... Make sure you use "xterm" or something similar.  Also, the `OutputStream` you give to the terminal needs to handle all the ansi sequences correctly, so it may also be a bit broken between this stream and the HTerm terminal.
